addons/binding/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpChannelState.java

```diff
@@ -0,0 +1,264 @@
+
+        this.stateUpdater = Optional.of(scheduler.scheduleWithFixedDelay(
+                () -> fetchState(stateRequest),
+                0,
+                stateRequest.getRefreshInterval().toMillis(),
+                TimeUnit.MILLISECONDS
+        ));
+    }
+
+    private void cancelStateFetch() {
+        this.stateUpdater = this.stateUpdater.flatMap(su -> {
+            su.cancel(false);
+            return Optional.empty();
+        });
+    }
+
+    private void fetchState(final HttpChannelConfig.StateRequest stateRequest) {
+        if (!this.fetchingState) {
+            this.fetchingState = true;
+            final URL url = stateRequest.getUrl();
+            makeHttpRequest(
+                    HttpChannelConfig.Method.GET,
+                    url,
+                    stateRequest.getConnectTimeout(),
+                    stateRequest.getRequestTimeout(),
+                    stateRequest.getUsername(),
+                    stateRequest.getPassword(),
+                    DEFAULT_CONTENT_TYPE,
+                    this.lastStateEtag,
+                    Optional.empty()
+            ).whenComplete((response, t) -> {
+                this.fetchingState = false;
+                if (t != null) {
+                    this.errorListener.accept(this.channelUID, ThingStatusDetail.COMMUNICATION_ERROR, "Connection to server failed when fetching state: " + t.getMessage());
+                } else if (response.getResponse().getStatus() / 100 != 2 && response.getResponse().getStatus() != HttpStatus.NOT_MODIFIED_304) {
+                    this.errorListener.accept(this.channelUID, ThingStatusDetail.COMMUNICATION_ERROR, "Server returned HTTP status " + response.getResponse().getStatus() + " when fetching state");
+                } else if (response.getResponse().getStatus() != HttpStatus.NOT_MODIFIED_304) {
+                    final State newState = stateFromResponse(response, stateRequest.getResponseTransform());
+                    logger.debug("[{}] got new state '{}'", this.channelUID.getId(), newState.toFullString());
+                    this.stateUpdatedListener.accept(this.channelUID, newState);
+                    this.lastStateEtag = Optional.ofNullable(response.getResponse().getHeaders().get("etag"));
+                }
+            });
+        }
+    }
+
+    private String doTransform(final Optional<Transform> maybeTransform, final String value) throws IllegalArgumentException {
+        return maybeTransform.map(transform -> transform.applyTransform(value)).orElse(value);
+    }
+
+    private State stateFromResponse(final HttpUtil.HttpResponse response, final Optional<Transform> transform) {
+        if (CHANNEL_TYPE_ID_IMAGE.equals(this.channelTypeUID.getId())) {
+            return response.asRawType();
+        } else {
+            try {
+                return TypeParser.parseState(STATE_TYPES, doTransform(transform, response.asString()));
+            } catch (final IllegalArgumentException e) {
```

Comment:
If you handle the exception of `doTransform` anyway, why not use the original TransformationException, instead of wrapping it in a IllegalArgumentException like you did. You can handle both exception types here with:
`IllegalArgumentException|TransformationException e`

Reply:
(I think you put this comment on the wrong line, but I see what you're getting at.)

So the issue is that `TransformationException` is a checked exception, so if I didn't transform it, I'd be unable to call it within a lambda in `doTransform()` to allow the exception to bubble out.  I agree that it makes sense to keep it separate, though, so instead of mapping on the Optional, I'll just do  isPresent+get instead and remove the wrapping.